File: apps/mobile/lib/features/workouts/data/workout_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/api/api_client.dart';
import '../../../core/api/api_error.dart';
import '../domain/workout.dart';

class WorkoutRepository {
  WorkoutRepository(this._dio);

  final Dio _dio;

  Future<WorkoutListPage> list({int limit = 20, String? cursor}) async {
    return _wrap(() async {
      final response = await _dio.get<Map<String, dynamic>>(
        '/workouts',
        queryParameters: {
          'limit': limit,
          'cursor': ?cursor,
        },
      );
      return WorkoutListPage.fromJson(response.data!);
    });
  }

  Future<Workout> get(String id) async {
    return _wrap(() async {
      final response = await _dio.get<Map<String, dynamic>>('/workouts/$id');
      return Workout.fromJson(response.data!);
    });
  }

  Future<Workout> start({String? name, String? notes, String? templateId}) {
    return _wrap(() async {
      final response = await _dio.post<Map<String, dynamic>>(
        '/workouts',
        data: {
          'name': ?name,
          'notes': ?notes,
          'templateId': ?templateId,
        },
      );
      return Workout.fromJson(response.data!);
    });
  }

  Future<WorkoutSet> addSet({
    required String workoutId,
    required String exerciseId,
    double? weightKg,
    int? reps,
    int? timeSeconds,
    int? distanceMeters,
    double? rpe,
  }) {
    return _wrap(() async {
      final response = await _dio.post<Map<String, dynamic>>(
        '/workouts/$workoutId/sets',
        data: {
          'exerciseId': exerciseId,
          'weightKg': ?weightKg,
          'reps': ?reps,
          'timeSeconds': ?timeSeconds,
          'distanceMeters': ?distanceMeters,
          'rpe': ?rpe,
        },
      );
      return WorkoutSet.fromJson(response.data!);
    });
  }

  Future<Workout> finish(String id, {String? notes}) {
    return _wrap(() async {
      final response = await _dio.post<Map<String, dynamic>>(
        '/workouts/$id/finish',
        data: {'notes': ?notes},
      );
      return Workout.fromJson(response.data!);
    });
  }

  Future<void> delete(String id) async {
    await _wrap(() => _dio.delete<void>('/workouts/$id'));
  }

  Future<T> _wrap<T>(Future<T> Function() call) async {
    try {
      return await call();
    } on DioException catch (error) {
      throw ApiError.fromDio(error);
    }
  }
}

final workoutRepositoryProvider = Provider<WorkoutRepository>((ref) {
  return WorkoutRepository(ref.watch(apiClientProvider).dio);
});

File: apps/api/_archive/repositories/workout.repository.ts
import type { Prisma, PrismaClient, Workout, WorkoutSet } from '@prisma/client';
import { prisma } from '../lib/prisma.js';

export interface ListWorkoutsArgs {
  userId: string;
  limit: number;
  cursor?: string;
}

export type WorkoutWithSets = Workout & { sets: WorkoutSet[] };

export class WorkoutRepository {
  constructor(private readonly db: PrismaClient = prisma) {}

  list({ userId, limit, cursor }: ListWorkoutsArgs): Promise<Workout[]> {
    return this.db.workout.findMany({
      where: { userId },
      orderBy: { startedAt: 'desc' },
      take: limit + 1,
      ...(cursor ? { skip: 1, cursor: { id: cursor } } : {}),
    });
  }

  findById(id: string, userId: string): Promise<WorkoutWithSets | null> {
    return this.db.workout.findFirst({
      where: { id, userId },
      include: { sets: { orderBy: { order: 'asc' } } },
    });
  }

  start(input: {
    userId: string;
    name?: string;
    templateId?: string;
    notes?: string;
  }): Promise<Workout> {
    return this.db.workout.create({
      data: {
        userId: input.userId,
        name: input.name,
        templateId: input.templateId,
        notes: input.notes,
      },
    });
  }

  update(
    id: string,
    userId: string,
    data: Prisma.WorkoutUpdateInput,
  ): Promise<Workout> {
    return this.db.workout.update({
      where: { id, userId },
      data,
    });
  }

  delete(id: string, userId: string): Promise<Workout> {
    return this.db.workout.delete({ where: { id, userId } });
  }

  async nextSetOrder(workoutId: string): Promise<number> {
    const last = await this.db.workoutSet.findFirst({
      where: { workoutId },
      orderBy: { order: 'desc' },
      select: { order: true },
    });
    return (last?.order ?? -1) + 1;
  }

  addSet(input: {
    workoutId: string;
    exerciseId: string;
    order: number;
    weightKg?: number | null;
    reps?: number | null;
    timeSeconds?: number | null;
    distanceMeters?: number | null;
    rpe?: number | null;
  }): Promise<WorkoutSet> {
    return this.db.workoutSet.create({
      data: {
        workoutId: input.workoutId,
        exerciseId: input.exerciseId,
        order: input.order,
        weightKg: input.weightKg ?? null,
        reps: input.reps ?? null,
        timeSeconds: input.timeSeconds ?? null,
        distanceMeters: input.distanceMeters ?? null,
        rpe: input.rpe ?? null,
        completedAt: new Date(),
      },
    });
  }
}

export const workoutRepository = new WorkoutRepository();
